akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/ProtobufProtocolConversions.scala

```diff
@@ -52,30 +53,51 @@ private[akka] object ProtobufProtocolConversions {
     offset match {
       case None => NoOffset
       case Some(o) =>
-        val timestamp =
-          o.timestamp.map(_.asJavaInstant).getOrElse(Instant.EPOCH)
-        val seen = o.seen.map {
-          case PersistenceIdSeqNr(pid, seqNr, _) =>
-            pid -> seqNr
-        }.toMap
-        TimestampOffset(timestamp, seen)
+        if (o.offsetsBySlice.nonEmpty) {
+          val offsets = o.offsetsBySlice.flatMap { offsetBySlice =>
+            offsetBySlice.offset.map(protocolOffsetToTimestampOffset).map(offsetBySlice.slice.->)
+          }
+          TimestampOffsetBySlice(offsets.toMap)
+        } else {
+          protocolOffsetToTimestampOffset(o)
+        }
     }
 
+  private def protocolOffsetToTimestampOffset(offset: proto.Offset): TimestampOffset = {
+    val timestamp =
+      offset.timestamp.map(_.asJavaInstant).getOrElse(Instant.EPOCH)
+    val seen = offset.seen.map {
+      case PersistenceIdSeqNr(pid, seqNr, _) =>
+        pid -> seqNr
+    }.toMap
+    TimestampOffset(timestamp, seen)
+  }
```

Comment:
Maybe worth optimizing this one, since it's used for each and every event?
Thinking of `match` instead of map + getOrElse, and special case for the normal seen with one element.

Reply:
Ok, just moved this as is. Can optimise.

Reply:
Updated in https://github.com/akka/akka-projection/pull/1181/commits/b7c8221abf29ae30e58419889c7796274546adc1